contracts/interchain-token-service/tests/flow_limit.rs

```diff
@@ -121,170 +202,228 @@ fn set_flow_limit_fails_invalid_amount() {
+            &data,
+            &gas_token,
+        ),
+        ContractError::FlowLimitExceeded
     );
 }
 
-#[test]
-#[should_panic(expected = "Error(Contract, #21)")] // ContractError::FlowAmountOverflow
-fn add_flow_fails_on_flow_comparison_overflow() {
-    let (env, client, gateway_client, _, signers) = setup_env();
-    register_chains(&env, &client);
-    let (token_id, sender) = setup_flow_limit(&env, &client);
-    let gas_token = setup_gas_token(&env, &sender);
+const LARGE_FLOW_LIMIT: i128 = i128::MAX - 50;
+const OVERFLOW_CASES: &[(i128, i128)] = &[
+    (LARGE_FLOW_LIMIT - 1, 2),
+    (i128::MAX - 100, 2),
+    (LARGE_FLOW_LIMIT, 1),
+];
 
-    client
-        .mock_all_auths()
-        .set_flow_limit(&token_id, &Some(i128::MAX - 50));
+enum OverflowTestDirection {
+    InThenOut,
+    OutThenIn,
+}
 
-    let high_amount = i128::MAX - 100;
-    let (source_chain, message_id, source_address, payload, messages) =
-        create_interchain_transfer_message(&env, &client, &token_id, high_amount);
-    approve_gateway_messages(&env, &gateway_client, signers, messages);
-    client.execute(&source_chain, &message_id, &source_address, &payload);
+fn test_overflow_case(case: &(i128, i128), direction: OverflowTestDirection) {
+    let (env, client, gateway, token) = setup();
+    let gas_token = setup_gas_token(&env, &token.deployer);
 
-    let small_amount = 100;
-    let destination_chain = String::from_str(&env, "ethereum");
-    let destination_address = Bytes::from_hex(&env, "4F4495243837681061C4743b74B3eEdf548D56A5");
+    client
+        .mock_all_auths()
+        .set_flow_limit(&token.id, &Some(LARGE_FLOW_LIMIT));
 
+    let (destination_chain, destination_address, data) = dummy_transfer_params(&env);
     client
         .mock_all_auths()
         .set_trusted_chain(&destination_chain);
 
-    client.mock_all_auths().interchain_transfer(
-        &sender,
-        &token_id,
-        &destination_chain,
-        &destination_address,
-        &small_amount,
-        &None,
-        &gas_token,
-    );
+    match direction {
+        OverflowTestDirection::InThenOut => {
+            let amount_in = case.0;
+            let amount_out = case.1;
+
+            let msg = approve_its_transfer(&env, &client, &gateway, &token.id, amount_in);
+
+            client.execute(
+                &msg.source_chain,
+                &msg.message_id,
+                &msg.source_address,
+                &msg.payload,
+            );
+
+            assert_contract_err!(
+                client.mock_all_auths().try_interchain_transfer(
+                    &token.deployer,
+                    &token.id,
+                    &destination_chain,
+                    &destination_address,
+                    &amount_out,
+                    &data,
+                    &gas_token
+                ),
+                ContractError::FlowAmountOverflow
+            );
+        }
+        OverflowTestDirection::OutThenIn => {
```

Comment:
just create two separate loops to go over each case instead of having a complicated match block in a helper